test/parallel/test-find-package-json.js

```diff
@@ -0,0 +1,155 @@
+
+
+describe('findPackageJSON', () => { // Throws when no arguments are provided
+  it('should throw when no arguments are provided', () => {
+    assert.throws(
+      () => findPackageJSON(),
+      { code: 'ERR_INVALID_ARG_TYPE' }
+    );
+  });
+
+  it('should throw when parentLocation is invalid', () => {
+    for (const invalid of ['', null, undefined, {}, [], Symbol(), 1, 0]) {
+      assert.throws(
+        () => findPackageJSON('', invalid),
+        { code: 'ERR_INVALID_ARG_TYPE' },
+        invalid
+      );
+    }
+  });
+
+  it('should accept a file URL (string), like from `import.meta.resolve()`', () => {
+    const importMetaUrl = `${pathToFileURL(fixtures.fixturesDir)}/whatever.ext`;
+    const specifierBase = './packages/root-types-field/';
+    assert.strictEqual(
+      findPackageJSON(`${specifierBase}index.js`, importMetaUrl),
+      fixtures.path(specifierBase, 'package.json')
+    );
+  });
+
+  it('should accept a file URL instance', () => {
+    const fixturesDirUrl = `${pathToFileURL(fixtures.fixturesDir)}`;
+    const importMetaUrl = `${fixturesDirUrl}/whatever.ext`;
+    const specifierBase = './packages/root-types-field/';
+    assert.strictEqual(
+      findPackageJSON(
+        new URL('index.js', `${fixturesDirUrl}${specifierBase.slice(1)}`),
+        new URL(importMetaUrl),
+      ),
+      fixtures.path(specifierBase, 'package.json'),
+    );
+  });
+
+  it('should be able to crawl up (CJS)', () => {
+    const pathToMod = fixtures.path('packages/nested/sub-pkg-cjs/index.js');
+    const parentPkg = require(pathToMod);
+
+    const pathToParent = fixtures.path('packages/nested/package.json');
+    assert.strictEqual(parentPkg, pathToParent);
+  });
+
+  it('should be able to crawl up (ESM)', () => {
+    const pathToMod = fixtures.path('packages/nested/sub-pkg-esm/index.js');
+    const parentPkg = require(pathToMod).default; // This test is a CJS file
+
+    const pathToParent = fixtures.path('packages/nested/package.json');
+    assert.strictEqual(parentPkg, pathToParent);
+  });
+
+  it('can require via package.json', () => {
+    const pathToMod = fixtures.path('packages/cjs-main-no-index/other.js');
+    // `require()` falls back to package.json values like "main" to resolve when there is no index
+    const answer = require(pathToMod);
+
+    assert.strictEqual(answer, 43);
+  });
+
+  it('should be able to do whatever the heck Antoine created', () => {
+    tmpdir.refresh();
+
+    fs.writeFileSync(tmpdir.resolve('entry.mjs'), `
+      import { findPackageJSON } from 'node:module';
+      import fs from 'node:fs';
+
+      const readPJSON = (locus) => JSON.parse(fs.readFileSync(locus, 'utf8'));
+
+      const { secretNumberSubfolder } = readPJSON(findPackageJSON('pkg', import.meta.url));
+      let secretNumberPkgRoot; // Impossible to resolve because pkg's pjson.exports blocks access
+      try { ({ secretNumberPkgRoot } = findPackageJSON('pkg/subfolder/', import.meta.url)) } catch {}
```

Comment:
This makes no sense, why are you trying to use `pkg/subfolder/` specifier to get the root `package.json`? Is this API useful if it still relies on `package.json` being made available by package authors? Like, why not rely on package authors exporting a `./package.json` leading to their `package.json`?

Reply:
Not every package author knows to do that, or is willing to do that, and tooling can't reliably get to package.json without some kind of utility in node core because the only paths they have to go on are the result of `exports` resolution, which might only terminate in a subfolder.

Reply:
Then shouldn't build an API that can actually access those unexported `package.json`? It seems to me the root `package.json` is likely much more useful than the subfolder one

Reply:
Sorry, to clarify, the thing that's critically important is the root package.json, but due to `type` and `main`, there's still lots of use cases for wanting to access a nested package.json.

Reply:
> This makes no sense, why are you trying to use pkg/subfolder/ specifier to get the root package.json?

Because I was tired and made a mistake renaming it

Reply:
We should document that this API is only useful to access nested `package.json`. But that doesn't seem very satisfying, I know some folks are defining `"exports"` map purposefully very limited, so they can change the internal file tree without affecting their users. As is, this API kinda break that, because if someone start relying on the non-presence of nested `package.json` to access some properties in the root, it means that introducing a nested `package.json` would now be a breaking change, even if both are not exported.

I think it's important we realize that:
- folks will use this API mainly to access the root `package.json` (I don't disagree that the nested one can also be useful, but arguably it's more niche)
- the current design is much too brittle to fulfil this usecase.

It feels to me introducing this API would be worth than the current status quo, and we should provide a solution for "finding the unexported root `package.json`" use-case (and IMO that's blocking, though my concern could be addressed with a plan and a bunch of TODOs).

Reply:
Let’s imagine pkg is released first with the following file structure:

- package.json
- subfolder/
  - index.mjs
- subfolder2/
  - index.mjs

Then the next version it’s changed to:

- package.json
- subfolder/
  - index.js
  - package.json
- subfolder2/
  - index.mjs
  - package.json

What was previously a non-breaking change (adding nested package.json) is now a breaking change. One the first version, `findPackageJSON` would return the root package.json, on the second version it would return the nested one.

Reply:
Ah, I see.

Do you have any idea how to sidestep this part of the resolution algorithm?

Reply:
adding a nested package.json is always a breaking change if it has a non-default `main` or `type`, which is the only reason to add one.

Reply:
It's observable by `require.resolve` of the path to the dir - also, not every package uses `exports`.

Reply:
My point is precisely that this API as-is is problematic because it breaks the assumption for packages that use `"exports"`, I'm of course aware that not every package uses `"exports"`